pkg/tools/serialize/serialize.go

```diff
@@ -39,13 +43,19 @@ func NewExecutor() Executor {
 		execCh:      make(chan func(), 100),
 		finalizedCh: make(chan struct{}),
 	}
-	go rv.eventLoop()
-	runtime.SetFinalizer(rv, func(f *executor) {
-		close(f.finalizedCh)
-	})
+	rv.wg.Add(1)
```

Comment:
Why the wait group?  Isn't the t.finalizeCh sufficient?

Reply:
Closing of finalizeCh is just a signal to stop without actual waiting for stop. At that time `executor` might hang in `exec()` call. So that's also more important for proper closing resource in tests, not for real life

Reply:
Thanks for shining a light on this.  Turns out that there was a bug in the finalization of serialize, which I've fixed in #83 (as well as adding good testing).  If the user-supplied function to be executed hangs forever, that is on them, not us.  We just need to test that we properly clean up on our end.

Reply:
Cool! In that case I revert finalizers changes

Reply:
Do go check the thing you were hitting before up in the heal.NewClient() finalizer... not sure whether it was related to the bug in serializer or the use of SetFinalizer here, and that's worth checking.

Reply:
Fixed with `per-chain` Context